### PR TITLE
New helpers features

### PR DESCRIPTION
**Pull request by RobGeada**
Added changes to data loader schema for pipeline

### CODE REVIEW

File: nutai/helpers.py [hunk 1]
```diff
@@ -5,85 +5,111 @@
...
+def tag_filter(x):
+    if 'Red Hat Enterprise Linux'in x:
+        x = re.sub(r'[\d]+[.]*[\d]+', '', x)
+    if x=='Red Hat JBoss Enterprise Application Platform':
+        return True
+    x = x.strip()
+    if x.count(" ")<5:
+        return True
+    return False
+
```
Comment (mattf) [inline nutai/helpers.py]: this should be done elsewhere. it's too specific to a dataset to include here.

File: nutai/helpers.py [hunk 2]
```diff
@@ -5,85 +5,111 @@
...
     else:
         return ids, texts
 
-
 # filename contents: csv file with three columns, id_a, id_b, label_ab, corresponding to testset duplicate pairs
 # docs contents: dict keyed by training set doc ids
 # returns lists of lists, each sublist of format [id_a (str), id_b (str), label_ab (int)]
-def load_testset(filename, docs):
-    with open(filename, "r") as f:
+def load_testset(filename, docs, verbose):
```
Comment (mattf) [inline nutai/helpers.py]: provide default to this so it doesn't change the function signature

File: nutai/helpers.py [hunk 3]
```diff
@@ -5,85 +5,111 @@
...
+    with open(filename,"r") as f:
```
Comment (mattf) [inline nutai/helpers.py]: run flake8. this needs a space after the comma.

File: nutai/helpers.py [hunk 4]
```diff
@@ -5,85 +5,111 @@
...
         testset = [line.split() for line in f.read().split("\n")[:-1]]
-    testset = [[pair[0], pair[1], int(pair[2])] for pair in testset if str(pair[0]) in docs and str(pair[1]) in docs]
-    print("Testset Size:", len(testset))
+    testset = [[pair[0],pair[1],int(pair[2])] for pair in testset if str(pair[0]) in docs and str(pair[1]) in docs]
```
Comment (mattf) [inline nutai/helpers.py]: pep8 violation

File: nutai/helpers.py [hunk 5]
```diff
@@ -5,85 +5,111 @@
...
+    if verbose:
+        print("Testset Size:",len(testset))
```
Comment (mattf) [inline nutai/helpers.py]: pep8 violation

File: nutai/helpers.py [hunk 6]
```diff
@@ -5,85 +5,111 @@
...
     return testset
-
+
```
Comment (mattf) [inline nutai/helpers.py]: remove whitespace from otherwise empty line

File: nutai/helpers.py [hunk 7]
```diff
@@ -5,85 +5,111 @@
...
+        print("Testset Size:",len(testset))
     return testset
-
+    
 
 def save_ids(ids):
     with open("ids", 'wb') as fp:
         msgpack.dump(ids, fp)
 
-
```
Comment (mattf) [inline nutai/helpers.py]: keep this extra line

File: nutai/helpers.py [hunk 8]
```diff
@@ -5,85 +5,111 @@
...
 def load_ids():
     with open("ids", 'rb') as fp:
         with Timer("load ids"):
             return msgpack.load(fp)
 
-
```
Comment (mattf) [inline nutai/helpers.py]: keep this extra line

File: nutai/helpers.py [hunk 9]
```diff
@@ -96,17 +122,14 @@ def save_scores(ids, score_generator):
         with open("scores", 'wb') as fp:
             msgpack.dump(scores, fp)
 
-
```
Comment (mattf) [inline nutai/helpers.py]: keep this extra line

File: nutai/helpers.py [hunk 10]
```diff
@@ -96,17 +122,14 @@ def save_scores(ids, score_generator):
...
 def load_scores():
     with open("scores", 'rb') as fp:
         with Timer("load scores"):
             return msgpack.load(fp)
 
-
```
Comment (mattf) [inline nutai/helpers.py]: keep this extra line

File: nutai/helpers.py [hunk 11]
```diff
@@ -96,17 +122,14 @@ def save_scores(ids, score_generator):
...
 def make_pair(id0, id1):
     return id0 < id1 and (id0, id1) or (id1, id0)
 
-
```
Comment (mattf) [inline nutai/helpers.py]: keep this extra line

File: nutai/helpers.py [hunk 12]
```diff
@@ -145,7 +168,6 @@ def load_tests(ids):
 
     return test_set, num_positive, num_negative
 
-
```
Comment (mattf) [inline nutai/helpers.py]: keep this extra line

File: nutai/helpers.py [hunk 13]
```diff
@@ -5,85 +5,111 @@
...
+                                     })
                     doc_tags = doc['tag'] if 'tag' in doc else []
                     doc_tags += doc['product'] if 'product' in doc else []
                     tags.append(doc_tags)
         ids = np.array(ids)
-    print("missing issues", missing_issue)
-    print("missing body", missing_body)
-    print("skipped solutions (missing both)", missing_both)
-    print(len(ids), ":", " ".join(map(str, ids[1:5])), "...", " ".join(map(str, ids[-4:])))
+        num_docs = len(ids)
```
Comment (mattf) [inline nutai/helpers.py]: remove unused num_docs

File: nutai/helpers.py [hunk 14]
```diff
@@ -5,85 +5,111 @@
...
+    if x.count(" ")<5:
+        return True
+    return False
+
 
 # filename contents: {[solution.id: "xyz", body: "text", issue: "text"]*}
 # extra_output: choose whether to return documents tags and raw strings as well
 # processor(string) -> (np.array of ids, list of [string*])
-def load_texts(filename, extra_output=False, preprocessing=simple_preprocess):
+def load_texts(filename, verbose, extra_output=False, preprocessing=simple_preprocess):
```
Comment (mattf) [inline nutai/helpers.py]: provide a default for verbose so it does not change function signature

File: nutai/helpers.py [hunk 15]
```diff
@@ -154,20 +176,17 @@ def load_tests(ids):
...
         del sims
-
-    sims = np.zeros((n_docs, n_docs), dtype=np.dtype('u1'))
-    for slice_idx in tqdm(range(0, n_docs, slice_size)):
-        sims[slice_idx:slice_idx+slice_size, :] = \
-            cosine_similarity(vect_mat[slice_idx:slice_idx+slice_size], vect_mat)*255
+    
+    sims = np.zeros((n_docs,n_docs),dtype=np.dtype('u1'))
+    for slice_idx in tqdm(range(0,n_docs,slice_size)):
+        sims[slice_idx:slice_idx+slice_size,:] = cosine_similarity(vect_mat[slice_idx:slice_idx+slice_size],vect_mat)*255
```
Comment (mattf) [inline nutai/helpers.py]: pep8 violations

File: nutai/helpers.py [hunk 16]
```diff
@@ -154,20 +176,17 @@ def load_tests(ids):
...
     return sims
 
 
-ConfusionMatrix = namedtuple('ConfusionMatrix', ['tn', 'fp', 'fn', 'tp'])
-
+ConfusionMatrix = namedtuple('ConfusionMatrix', ['tn','fp','fn','tp'])
```
Comment (mattf) [inline nutai/helpers.py]: pep8 violation

File: nutai/helpers.py [hunk 17]
```diff
@@ -154,20 +176,17 @@ def load_tests(ids):
...
 
 def get_score_from_matrix(scores, ids, id0, id1):
-    return scores[ids == id0, ids == id1][0]
-
+    return scores[ids==id0,ids==id1][0]
```
Comment (mattf) [inline nutai/helpers.py]: pep8 violation

File: nutai/helpers.py [hunk 18]
```diff
@@ -176,8 +195,7 @@ def get_score_from_docvecs(docvecs, ids, id0, id1):
     else:
         return score
 
-
 def evaluate(scores, test_set, ids, threshold, get_score=get_score_from_matrix):
     y_true = list(test_set.values())
     y_pred = [get_score(scores, ids, id0, id1) > threshold for id0, id1 in test_set]
-    return ConfusionMatrix(*confusion_matrix(y_true, y_pred).ravel())
+    return ConfusionMatrix(*confusion_matrix(y_true, y_pred).ravel())
```
Comment (mattf) [inline nutai/helpers.py]: missing newline at end of file

File: nutai/helpers.py (hunk 16 above)
Comment (RobGeada) [reply nutai/helpers.py]: I didn't write this line?

**Pull request merged**